### DISCUSSION TITLE
Could you give an example of implementing a two-dimensional Fourier transform by use torchkbnufft?

### DISCUSSION THREAD

**Discussion by lalalashini** [mmuckley/torchkbnufft#82] [Q&A]
Hello,torchkbnufft is great but the example is not easy.So,I want to know if I want to  compute the Fourier transform of a two-dimensional array by use torchkbnufft,what should I do?
example:a = torch.rand((4,4))
          fft: b = torch.fft.fft2(a)
torchkbnufft:???

**Comment by mmuckley** [comment]
Converting this to a Discussion as it does not document a bug.

At a high level the equivalent is

```python
import torchkbnufft as tkbn

a = torch.rand(4, 4)
nufft_ob = tkbn.KbNufft(im_size=(4, 4))
b = nufft_ob(a, ktraj)
```

You need to define a `ktraj` or k-space trajectory that defines the Fourier points you want to sample. You can think of a regular FFT as estimating the points on a grid between -pi and pi, so you can define your off-grid points in the `ktraj` with that intuition.

For a full detailed example for 2D arrays, please see the [Basic Example notebook](https://colab.research.google.com/github/mmuckley/torchkbnufft/blob/main/notebooks/Basic%20Example.ipynb).

**Comment by lalalashini** [comment]
import torch
import torchkbnufft as tkbn

a = torch.rand(4, 4)+1j
print(torch.fft.fft2(a))


nufft_ob = tkbn.KbNufft(im_size=(4, 4))

ktraj = torch.ones(4,4)*torch.pi
print(ktraj)
b = nufft_ob(a, ktraj)

print(b)


D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\_nufft\interp.py:385: UserWarning: __rfloordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  if omega.shape[1] / (num_threads // factor) >= min_kspace_per_fork:
Traceback (most recent call last):
  File "D:\pycharm\paper\nufftpy-master1\nufftpy\5.py", line 12, in <module>
    b = nufft_ob(a, ktraj)
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torch\nn\modules\module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\modules\kbnufft.py", line 211, in forward
    output = tkbnF.kb_table_nufft(
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\functional\nufft.py", line 172, in kb_table_nufft
    data = kb_table_interp(
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\functional\interp.py", line 117, in kb_table_interp
    data = KbTableInterpForward.apply(
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\_autograd\interp.py", line 87, in forward
    output = table_interp(
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\_nufft\interp.py", line 392, in table_interp
    kdat = table_interp_fork_over_kspace(
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\_nufft\interp.py", line 310, in table_interp_fork_over_kspace
    # collect the results
    for ind, future in enumerate(futures):
        kdat[:, :, ind:klength:num_forks] = torch.jit.wait(future)
                                            ~~~~~~~~~~~~~~ <--- HERE

    return kdat
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\_nufft\interp.py", line 288, in <forked function>
    for omega_chunk in omega_chunks:
        futures.append(
            torch.jit.fork(
            ~~~~~~~~~~~~~~~
                table_interp_one_batch,
                ~~~~~~~~~~~~~~~~~~~~~~~
                image,
                ~~~~~~
                omega_chunk,
                ~~~~~~~~~~~~
                tables,
                ~~~~~~~
                n_shift,
                ~~~~~~~~
                numpoints,
                ~~~~~~~~~~
                table_oversamp,
                ~~~~~~~~~~~~~~~
                offsets,
                ~~~~~~~ <--- HERE
            )
        )
  File "D:\anaconda\envs\pytorchcuda\lib\site-packages\torchkbnufft\_nufft\interp.py", line 171, in table_interp_one_batch

    # convert to normalized freq locs
    tm = omega / (2 * np.pi / grid_size.to(omega).unsqueeze(-1))
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE

    # compute interpolation centers
RuntimeError: The size of tensor a (4) must match the size of tensor b (0) at non-singleton dimension 0

WHY?

**Comment by mmuckley** [comment]
Sorry, I didn't test my example that I wrote. We need batch and channel dimensions (a PyTorch convention). The notebook should be the primary source of truth. Please consult the notebooks for the future.

```python
import torch
import torchkbnufft as tkbn

nufft_ob = tkbn.KbNufft(im_size=(4, 4))
a = torch.rand(1, 1, 4, 4) + 1j*torch.rand(1, 1, 4, 4)
ktraj = torch.randn(2, 128)
b = nufft_ob(a, ktraj)
```
